File: 18-Command/src/go/remote_control.go
package main

import "fmt"

// RemoteControl is the Invoker — holds a command, triggers it, and records
// history for undo. It knows nothing about receivers.
type RemoteControl struct {
	slot    Command
	history []Command
}

func (r *RemoteControl) SetCommand(command Command) {
	r.slot = command
}

func (r *RemoteControl) PressButton() {
	if r.slot != nil {
		r.slot.Execute()
		r.history = append(r.history, r.slot)
	}
}

func (r *RemoteControl) PressUndo() {
	if len(r.history) == 0 {
		fmt.Println("  [undo] nothing to undo")
		return
	}
	fmt.Println("  [undo]")
	last := r.history[len(r.history)-1]
	r.history = r.history[:len(r.history)-1]
	last.Undo()
}

File: 18-Command/src/rust/remote_control.rs
use std::rc::Rc;

use crate::command::Command;

/// Invoker — holds a command, triggers it, and records history for undo.
/// It knows nothing about receivers, only about `Command`.
pub struct RemoteControl {
    slot: Option<Rc<dyn Command>>,
    history: Vec<Rc<dyn Command>>,
}

impl RemoteControl {
    pub fn new() -> Self {
        RemoteControl { slot: None, history: Vec::new() }
    }

    pub fn set_command(&mut self, command: Rc<dyn Command>) {
        self.slot = Some(command);
    }

    pub fn press_button(&mut self) {
        if let Some(command) = &self.slot {
            command.execute();
            self.history.push(Rc::clone(command));
        }
    }

    pub fn press_undo(&mut self) {
        match self.history.pop() {
            Some(command) => {
                println!("  [undo]");
                command.undo();
            }
            None => println!("  [undo] nothing to undo"),
        }
    }
}
